src/Open3D/Core/Kernel/ReductionCUDA.cu

```diff
@@ -0,0 +1,1101 @@
+
+        // Reducing layers of function calls so compiler could do proper loop
+        // unroll that exposes instruction level parallelism.
+        while (idx < config_.num_inputs_per_output_) {
+            // load input
+            SmallArray<scalar_t, vt0> values;
+            if (input_calc_.dims_ == 1) {
+                StridedIterate<vt0>(
+                        [&](index_t i, index_t idx) {
+                            values[i] = data[idx * element_stride];
+                        },
+                        idx, end, stride);
+            } else {
+                StridedIterate<vt0>(
+                        [&](index_t i, index_t idx) {
+                            values[i] = data[input_calc_.get(idx)[0] /
+                                             sizeof(scalar_t)];
+                        },
+                        idx, end, stride);
+            }
+            // compute
+            StridedIterate<vt0, index_t>(
+                    [&](index_t i, index_t idx) {
+                        value_list[i] =
+                                ops_.Reduce(value_list[i], values[i], idx);
+                    },
+                    idx, config_.num_inputs_per_output_, config_.step_input_);
+            // step offset
+            idx += config_.step_input_ * vt0;
+        }
+#pragma unroll
+        for (int i = 1; i < vt0; i++) {
+            value_list[0] = ops_.Combine(value_list[0], value_list[i]);
+        }
+        return value_list[0];
+    }
+
+    OPEN3D_DEVICE arg_t BlockXReduce(arg_t value, char* shared_memory) const {
+        int dim_x = blockDim.x;
+        arg_t* shared = (arg_t*)shared_memory;
+        if (dim_x > warpSize) {
+            int address_base = threadIdx.x + threadIdx.y * blockDim.x;
+            shared[address_base] = value;
+            for (int offset = dim_x / 2; offset >= warpSize; offset >>= 1) {
+                __syncthreads();
+                if (threadIdx.x < offset && threadIdx.x + offset < blockDim.x) {
+                    arg_t other = shared[address_base + offset];
+                    value = ops_.Combine(value, other);
+                    shared[address_base] = value;
+                }
+            }
+            dim_x = warpSize;
+        }
+
+        __syncthreads();
+
+        for (int offset = 1; offset < dim_x; offset <<= 1) {
+            arg_t other = ops_.WarpShflDown(value, offset);
+            value = ops_.Combine(value, other);
+        }
+        return value;
+    }
+
+    OPEN3D_DEVICE arg_t BlockYReduce(arg_t value, char* shared_memory) const {
+        arg_t* shared = (arg_t*)shared_memory;
+        shared[config_.SharedMemoryOffset(0)] = value;
+        for (int offset = blockDim.y / 2; offset > 0; offset >>= 1) {
+            __syncthreads();
+            if (threadIdx.y < offset && threadIdx.y + offset < blockDim.y) {
+                arg_t other = shared[config_.SharedMemoryOffset(offset)];
+                value = ops_.Combine(value, other);
```

Comment:
Similar to the above, since `threadIdx.y < offset`, we can show that `shared[threadIdx.x + (threadIdx.y + offset) * blockDim.x]` is always after `shared[threadIdx.x + threadIdx.y * blockDim.x]` for all possible threads in the block.